content/news/2021/2021-09-20-associate-users-with-events-to-track-event-impact.md

```diff
@@ -0,0 +1,251 @@
+
+```js
+app.listen(4000, () => {
+  console.log("Server is listening on port 4000");
+})
+```
+
+Then start the server: 
+
+```
+npm run start
+```
+
+Then in another command line tab, run: 
+
+```
+curl --location --request GET 'http://localhost:4000/users'
+```
+
+You should get an unauthorized response which means the middleware is doing its job. Let's add in an API Key: 
+
+```
+curl --location --request GET 'http://localhost:4000/users' \
+--header 'x-api-key: 123'
+```
+
+Now, you should get back an array of two users with their API keys removed. Perfect! Now we can get started on what we came here to do—associate user data to events tracked in Exceptionless. 
+
+Create a new endpoint in that same file that looks like this: 
+
+```js
+router.post('/', authentication, function(req, res, next) {
+  try {
+    if(!validBody(req.body)) {
+      throw "No user object provided";
+    }
+    users.push({
+      apiKey: null, 
+      user: req.body
+    });
+    res.json(req.body);
+  } catch (error) {
+    res.status(400).send(error);
+  }
+})
+```
+
+we also need to create our `validBody` function. You can add this right below the `authentication` middleware function: 
+
+```js
+const validBody = (body) => {
+  if(!body.name || !body.email || !body.userId) {
+    return false;
+  }
+
+  return true;
+}
+```
+
+This endpoint allows anyone with a valid API key to add new users. New users should be added without an API. If the request body is empty, we will throw an error. Before wiring this up to Exceptionless, let's test this endpoint out. 
+
+But wait, I don't want to have to keep restarting the server when we make changes. Let's make this easier. Kill your server, then run: 
+
+```
+npm i nodemon
+```
+
+In your `pacakage.json` file, add a `dev` script like this: 
+
+```
+"dev": "nodemon node ./bin/www"
+```
+
+Now, restart your server with `npm run dev`. After any changes to the code, the server will automatically restart. 
+
+Ok, now let's test our endpoint without and with a body payload: 
+
+```
+curl --location --request POST 'http://localhost:4000/users' \
+--header 'x-api-key: 123'
+```
+
+You should get an error back. Now, try with a valid body: 
+
+```
+curl --location --request POST 'http://localhost:4000/users' \
+--header 'x-api-key: 123' \
+--header 'Content-Type: application/json' \
+--data-raw '{
+    "name": "New Guy", 
+    "email": "newguy@email.com",
+    "userId": "97f78146-3967-4a80-972c-2fcccad6f45f"
+}'
+```
+
+Now, you should get a 200 response with the newly added resource. Ok, we have our endpoint working. Let's connect it to Exceptionless so that we can track events by user. 
+
+At the top of your file, require in Exceptionless: 
+
+```js
+import { Exceptionless } from "@exceptionless/node";
+Exceptionless.startup("YOUR API Key");
```

Comment:
these exceptionless calls should be awaited

Reply:
I remember @ejsmith suggesting that we didn't need all these awaits at one point, but maybe he was suggesting the code needed to be refactored rather than it not being necessary. FWIW, this code works fine but there might be edge cases where it would fail.

Reply:
I think it's best to be consistent with our examples and docs. It's not technically needed but it should be awaited.